api/group_creation/group_manager.py

```diff
@@ -0,0 +1,150 @@
+# DO NOT USE IN PROD, WORK IN PROGRESS
+
+from collections.abc import Iterable
+from dataclasses import dataclass
+from datetime import datetime
+
+from django.utils import timezone
+
+from api.models import AgeRange, Group, Student, Teacher
+from api.models.auxil.constants import (
+    DEFAULT_LESSON_DURATION_MIN,
+    MAX_AGE_KIDS_GROUP,
+    MAX_AGE_TEEN_GROUP,
+)
+from api.models.auxil.status_setter import StatusSetter
+from api.models.choices.communication_language_mode import CommunicationLanguageMode
+from api.models.choices.log_event_type import (
+    GroupLogEventType,
+    StudentLogEventType,
+    TeacherLogEventType,
+)
+from api.models.choices.status import GroupStatus, TeacherStatus
+from api.models.language_and_level import LanguageAndLevel
+from api.processors.auxil.log_event_creator import GroupLogEventCreator
+
+
+class GroupManager:
+    @dataclass
+    class GroupSizeRestriction:
+        min_number: int
+        max_number: int
+
+    @dataclass
+    class GroupCandidate:
+        language_and_level: LanguageAndLevel
+        communication_language_mode: CommunicationLanguageMode
+        age_range: AgeRange
+        teacher: Teacher
+        students: list[Student]
+        monday: datetime | None = None
+        tuesday: datetime | None = None
+        wednesday: datetime | None = None
+        thursday: datetime | None = None
+        friday: datetime | None = None
+        saturday: datetime | None = None
+        sunday: datetime | None = None
+
+        def is_acceptable(self) -> bool:
+            # Checks group size is appropriate for age range
+            restrictions = GroupManager._get_allowed_group_size(self.age_range)
+            students_num = len(self.students)
+            if restrictions.min_number > students_num:
+                return False
+            if restrictions.max_number < students_num:
+                raise ValueError(
+                    f"Group candidate must respect upper group boundaries, "
+                    f"but {students_num} > {restrictions.max_number}"
+                )
+            return True
```

Comment:
That is not what I meant exactly. A suggested group with too few students is a "normal thing", which means "we couldn't find enough students with the higher priority, so we are not creating this group and go further".

If group suggestion code returned more than maximum allowed number of students, we should fail early because it means that preceding logic is wrong. The code sorting students and limiting their number should take it all into account at the same time, so that's just an assertion.

That particular assertion (on max limit) could happen in post-init method instead, that's true. But this sounds like a matter of personal preference to me.

Reply:
> That is not what I meant exactly. A suggested group with too few students is a "normal thing", which means "we couldn't find enough students with the higher priority, so we are not creating this group and go further".

OK, I think I understand what you mean. But on the other hand, if our algorithm tries to create a group with too few students, how is that fundamentally different from creating a group with too many? Maybe the algorithm shouldn't even attempt to create a group in that case? Or maybe, following the Pythonic fashion of using `try...except...` instead of `if...else...`, maybe the algorithm should try to create a group with too few students or too many and catch the exception?

It's just that when I look "from the outside", I can barely see the reason why the one is normal case, while the other one is a coding error. I do see it after thinking about it for a couple of minutes :)

By the way, when I think of it, it seems that my suggested `.is_acceptable` is completely redundant. What if we scrap that method altogether and just try and create a group and if it's not acceptable, it throws an error in `.post_init`, then the algorithm catches it and goes on without even calling the `.is_acceptable` method?

Reply:
> That particular assertion (on max limit) could happen in post-init method instead, that's true. But this sounds like a matter of personal preference to me.

Yes, definitelu a matter of personal preference. I suggested separating it to keep the distinction between a coding error (that should trigger an exception as early as possible, even before calling any methods) and a meaningful `False`.

Reply:
I guess my general question is - why don't you want to just throw exceptions in any case of limits not being observed? To me, it will make the logic more consice and predictable. But I may be missing something.

Reply:
I also prefer consistency, and from my point of view - there are no such big difference between any case where limit is violated

Reply:
The logic I had in mind does have have a major difference. The group candidate generated by different database filters can easily return less than minimum people and it's alright, no exception needed. We just don't pick this group while iterating candidates in the order of priority. Filters should be smart enough to put maximum restriction in place before returning the "candidate", so violating this is an inconsistency

Reply:
But I can put this extra check inside the main generation logic and not here. Then validation will look different, as you both see fit. I don't want to argue on it, so let's agree and move forward :)